xarray/backends/rasterio_.py

```diff
@@ -41,8 +41,10 @@ def shape(self):
         return self._shape
 
     def __getitem__(self, key):
-        key = indexing.unwrap_explicit_indexer(
-            key, self, allow=(indexing.BasicIndexer, indexing.OuterIndexer))
+        # TODO support indexing.decompose_indexer
```

Comment:
I think I need someone's help to update indexing scheme for rasterio backend.

Reply:
rasterio is strictly a slicing approach for fetching data, or pulling out individual data points. In this case, you want to avoid converting the slices into numpy arrays of indexes. Maybe it would make sense to have a function that could re-compose numpy arrays of indexes into groups of slices? So, if we were indexing `[1, 3, 4, 5, 2, 6]`, we would get back a `slice(1, 7)`? Maybe even make it smart enough to take the chunk hints so that indexes like `[1, 4, 5]` is kept as `slice(1, 6)`, rather than `(slice(1, 2), slice(4, 6))`?

Reply:
@WeatherGod Thanks for the review.
> Maybe it would make sense to have a function that could re-compose numpy arrays of indexes into groups of slices?

Agreed, it would be an improvement. But I am wondering the combination will explode if the dimension increases. Some heuristics would be also necessary how many chunks we used. 
I do not have yet a clear solution. I am open to any idea for the improvement.

Reply:
Yeah, we certainly have time to think about it. This feature doesn't seem to change anything for indexing that worked before, so I don't think we have to worry about regressions. Improvements to indexing can be thought of holistically as a separate task.

For our immediate needs though, for rasterio, we probably should just create a failback indexer that would just load up that entire dimension and index it as given if the original indexer is anything but scalar integers or slices.

Reply:
@fmaussion can you help me for this? 
What I want to do here is the 2-staged indexing.
1. index backend array by the 1st indexer.
2. index the loaded np.ndarray by the 2nd vectorized indexer.

These two indexers (for backend and np.ndarray) can be computed by
```python
key, np_inds = indexing.decompose_indexer(
    key, self.shape, indexing.IndexingSupport.OUTER)
```

I am not familiar with rasterio data structure (I do not yet understand what is going on below...), 
I would appreciate if you could instruct me.

Reply:
Rasterio (as wrapped by xarray) supports a strange form of indexing:
- Outer indexing for the first indexer element only.
- Basic indexing for remaining two indexer elements (every rasterio array is 3-dimensional).

I think the simplest fix would be to extend our rasterio wrapper to support arbitrary outer indexing.  This would entail changing the final clause below, where rasterio currently raises an error:
https://github.com/pydata/xarray/blob/e0621c7d66c13b486b1890f67a126caec2990da7/xarray/backends/rasterio_.py#L72-L78

Instead, we would want to create a slice object based on the minimum/maximum elements, and then add another indexing to `np_inds` to pull out the listed elements (after subtracting the `start` from the slice).